parlai/chat_service/core/chat_service_manager.py

```diff
@@ -0,0 +1,677 @@
+
+
+class ChatServiceManager(ABC):
+    class ChatServiceMessageSender(ABC):
+        """ChatServiceMessageSender is a wrapper around requests that simplifies the
+        the process of sending content.
+        """
+
+        @abstractmethod
+        def send_read(self, receiver_id):
+            pass
+
+        @abstractmethod
+        def typing_on(self, receiver_id, persona_id=None):
+            pass
+
+    def __init__(self, opt):
+        """Create a ChatServiceManager using the given setup options."""
+        # Manager attributes
+        self.opt = opt
+        self.server_url = None
+        self.port = 443
+        self.agent_pool_change_condition = threading.Condition()
+        self.active_worlds = {}
+        self.socket = None
+        self.sender = None
+        self.running = True
+        self.conversation_index = 0
+        self.shutting_down = False
+        self.bypass_server_setup = self.opt.get('bypass_server_setup')
+
+        # Messaging interaction functions that determine what to do when
+        # messages are confirmed as delivered, marked as read by a user, and
+        # noted as read by the bot.
+        self.confirm_message_delivery = self._confirm_message_delivery
+        self.handle_message_read = self._handle_message_read
+        self.handle_bot_read = self._handle_bot_read
+
+    def _parse_config(self, opt):
+        """
+        Parse config for task. Use this to parse all options and settings
+        necessary to set the variables for the conversation
+        """
+
+        self.config = opt['config']
+        self.overworld = self.config['overworld']
+        self.world_path = self.config['world_path']
+        self.world_module = shared_utils.get_world_module(self.world_path)
+        self.task_configs = self.config['configs']
+        self.max_workers = self.config['max_workers']
+        self.opt['task'] = self.config['task_name']
+        self.world_runner = MessengerWorldRunner(
+            opt, self.world_path, self.max_workers, self, opt['is_debug']
+        )  # Replace with base runner
+        self.max_agents_for = {
+            task: cfg.agents_required for task, cfg in self.task_configs.items()
+        }
+        self.onboard_map = {
+            task: cfg.onboarding_name for task, cfg in self.task_configs.items()
+        }
+        self.taskworld_map = {
+            task: cfg.task_name for task, cfg in self.task_configs.items()
+        }
+
+    @abstractmethod
+    def parse_additional_args(self, opt):
+        """Parse any other service specific args here."""
+        # page id for messenger to be obtained here
+
+    def _get_port(self):
+        """Return the port number currently being used."""
+        return self.port
+
+    def _set_port(self, port_no):
+        """
+        Use a custom port number
+        :param port_no: New port number to be used
+        """
+        self.port = port_no
+
+    @abstractmethod
+    def _complete_setup(self):
```

Comment:
should this function be called somewhere, e.g. in `__init__`?

Reply:
This is how it was defined for messenger and it wasn't called from __init__. The purpose of this function is that it is a sort of ```main``` from where all setup would be called. Not sure where it is called, any thoughts ?

Reply:
Ahh, i see. That is fine - could you please specify in the docstring that it would be called upon instantiation?